### PR TITLE
setup pagination for chat room

### LINKED ISSUE #936: [FEATURE] Set up pagination for chat
### Description

Initially chat just show last 20 message. I want to set up pagination for it such that whenever someone scroll to top they get more messages loaded. And also the chat page starts from top and not from bottom like any other chat. I want to fix that also.

### Screenshots

_No response_

### Additional information

_No response_

### Record

- [X] I agree to follow this project's Code of Conduct
- [X] I have checked the existing issues
- [X] I'm a GSSoC'23 contributor
- [X] I want to work on this issue

### PR DESCRIPTION
**Pull request by Palakkgoyal**
This PR closes #936 
I have added pagination on the chat room and also made it scroll from bottom to top.

### CODE REVIEW

**Review by narayan954 — CHANGES_REQUESTED** [review]
could this usage be due to this?? If so this could be alarming actually.
image

Also, there're many snapshots in our project which are not unsubscribed/closed. We need to lookout again for the db usage. I think before this chat feature, usage was fine...If we allow this pagination, it'd cross the limit very early tomorrow as there'd be more users. 


As now I am thinking, maybe this usage is partly because the snapshot is not closed and chat gets updated in background of users,once he has entered the chat section.It's like opening a portal lol on opening chat page. and partly because the pagination loads the chat very aggressively.

We need to add unsubscribe function in the return() of these component lifecycle.
I was just going to read [this article](https://blog.logrocket.com/how-to-use-react-hooks-firebase-firestore/) regarding firebase today only, maybe you can give it a read as well :)

**Comment by narayan954** [comment]
I think adding unsubscribe method should work, and this pagination is fine as well. If we get into a problem with this though, we can add pagination to be not on scroll but a button `load more`

**Comment by Palakkgoyal** [comment]
> I think adding unsubscribe method should work, and this pagination is fine as well. If we get into a problem with this though, we can add pagination to be not on scroll but a button `load more`

I knew that providing a cleanup function is important but due to lack of experience I was not able to grasp its importance. After watching this and reading more about it I understand why it is important. I will do the required changes in it. Should I add a button now for loading chats or not?

**Comment by narayan954** [comment]
> > I think adding unsubscribe method should work, and this pagination is fine as well. If we get into a problem with this though, we can add pagination to be not on scroll but a button `load more`
> 
> I knew that providing a cleanup function is important but due to lack of experience I was not able to grasp its importance. After watching this and reading more about it I understand why it is important. I will do the required changes in it. Should I add a button now for loading chats or not?

it's your call, if it doesn't seem like it, let it be,else you definitely can! :)

**Comment by Palakkgoyal** [comment]
I have done the required changes. Can you please check.

**Comment by narayan954** [comment]
> I have done the required changes. Can you please check.

aight, on it!

File: src/components/CommunityChat/index.jsx [hunk 1]
```diff
@@ -56,8 +70,51 @@ const ChatBox = () => {
         setMessages(data);
       });
 
-    return unsubscribe;
-  }, [db]); // TODO - won't this cause infinite calling/refreshing if the db is under use / always changing?
```
Comment (narayan954) [inline src/components/CommunityChat/index.jsx:60]: now after removing, I'm thinking , it wont be a live chat, right?

**Review by narayan954 — APPROVED** [review]
looks good to me!

**Comment by narayan954** [comment]
Wait, does it work at all @Palakkgoyal ;-; I missed to check and now on production, I don't think it works ;-;

**Comment by Palakkgoyal** [comment]
I checked it worked but I don't know about the production.

**Comment by narayan954** [comment]
> I checked it worked but I don't know about the production.

yes it works there, but doesn't at all in production version.

**Comment by Palakkgoyal** [comment]
I tried to give a good cleanup to it using AbortController but don't know why it was not working.

**Comment by narayan954** [comment]
> I tried to give a good cleanup to it using AbortController but don't know why it was not working.

now I think abortcontroller is what's causing it to abort.

**Comment by Palakkgoyal** [comment]
But I didn't used it cause AbortController wasn't even working during development.

**Comment by narayan954** [comment]
> But I didn't used it cause AbortController wasn't even working during development.

lol, btw in mobile devices, the feature work, but not in desktops... I'll go through the code again and try to fix as well

**Comment by Palakkgoyal** [comment]
> > But I didn't used it cause AbortController wasn't even working during development.
> 
> lol, btw in mobile devices, the feature work, but not in desktops... I'll go through the code again and try to fix as well

Is it? That's seems typical.

**Comment by narayan954** [comment]
> > > But I didn't used it cause AbortController wasn't even working during development.
> > 
> > 
> > lol, btw in mobile devices, the feature work, but not in desktops... I'll go through the code again and try to fix as well
> 
> Is it? That seems typical.

I feel the same, but it worked ;-;

**Comment by Palakkgoyal** [comment]
That's amazing.

**Pull request merged**